Tests/E2E/MacTest.cs

```diff
@@ -204,5 +204,139 @@ private async Task Test_OTAA_Unconfirmed_Send_And_Receive_C2D_Mac_CommandsImplAs
             Assert.True(foundMacCommandReceivedMsg, $"Did not find in network server logs: '{macCommandReceivedMsg}'");
             Assert.True(foundDeviceMacCommandResponseMsg, $"Did not find in network server logs: '{deviceMacCommandResponseMsg}'");
         }
+
+        /// <summary>
+        /// Ensures that data rate is updated on a device after a LinkADRCmd MAC Command is sent as C2D message.
+        /// </summary>
+        [RetryFact]
+        public async Task Data_Rate_Is_Updated_When_C2D_With_LinkADRCmd_Received()
+        {
+            const int messagesToSend = 10;
+            const int warmUpMessageCount = 2;
+
+            var device = TestFixtureCi.Device32_ABP;
+            LogTestStart(device);
+
+            // Setup LoRa device properties
+            await ArduinoDevice.setDeviceModeAsync(LoRaArduinoSerial._device_mode_t.LWABP);
+            await ArduinoDevice.setIdAsync(device.DevAddr, device.DeviceID, device.AppEUI);
+            await ArduinoDevice.setKeyAsync(device.NwkSKey, device.AppSKey, device.AppKey);
+
+            // Setup protocol properties
+            // Start with DR5
+            await ArduinoDevice.SetupLora(TestFixtureCi.Configuration.LoraRegion, LoRaArduinoSerial._data_rate_t.DR5, 4, true);
+            await TestFixture.CleanupC2DDeviceQueueAsync(device.DeviceID);
+
+            // Sends 2x unconfirmed messages
+            for (var i = 1; i <= warmUpMessageCount; ++i)
+            {
+                var msg = PayloadGenerator.Next().ToString(CultureInfo.InvariantCulture);
+                Log($"{device.DeviceID}: Sending unconfirmed '{msg}' {i}/{messagesToSend}");
+                await ArduinoDevice.transferPacketAsync(msg, 10);
+
+                await Task.Delay(Constants.DELAY_BETWEEN_MESSAGES);
+                await AssertUtils.ContainsWithRetriesAsync("+MSG: Done", ArduinoDevice.SerialLogs);
+                TestFixture.ClearLogs();
+            }
+
+            var c2dMessage = new LoRaCloudToDeviceMessage()
+            {
+                Fport = 0,
+                Payload = string.Empty,
+                MacCommands = { new LinkADRRequest(3, 4, 25, 0, 1) } // Update data rate to DR3
+            };
+
+            await TestFixtureCi.SendCloudToDeviceMessageAsync(device.DeviceID, c2dMessage);
+            Log($"C2D Message sent to device, need to check if it receives");
+
+            var foundC2DMessageCount = 0;
+            var foundLinkADRCmdCount = 0;
+            var foundChangedDataRateCount = 0;
+
+            var c2dLogMessage = $"{device.DeviceID}: cloud to device MAC command LinkADRCmd received Type: LinkADRCmd Answer, datarate: 3";
+            var linkADRCmdMessage = "LinkADRCmd mac command detected in upstream payload";
+            var changedDataRateMessage = $"\"datr\":\"SF9BW125\""; // DR3
```

Comment:
Which code part do we assert that is hit by checking for this message? Either a comment or a larger message (such that we can find the log statement where this is coming from) would make it easier to see what is asserted.

Reply:
The message will look something like
` 0100000000000032: sending message {"time":"12103424213013044","tmms":0,"tmst":214454836,"freq":867.1,"chan":3,"rfch":0,"stat":1,"modu":"LORA","datr":"SF9BW125","codr":"4/5","rssi":-113,"lsnr":6.0,"size":0,"data":{"value":104},"port":1,"fcnt":3,"rawdata":"MTA0","eui":"0100000000000032","gatewayid":"itestarm1","edgets":1638869280975} to hub`
Because there are many other values I think it's hard to assert on the entire log line and I didn't see other logs in the LNS log that would confirm the used data rate is what's expected.